diff --git a/src/std-data/model.cmake b/src/std-data/model.cmake
index e9e7099..0c49e5e 100644
--- a/src/std-data/model.cmake
+++ b/src/std-data/model.cmake
@@ -34,6 +34,18 @@ register_flag_optional(USE_ONEDPL
           DPCPP  - Implements policies through SYCL2020.
                    This requires the DPC++ compiler (other SYCL compilers are untested), required SYCL flags are added automatically."
         "OFF")
+register_flag_optional(CLANG_OFFLOAD
+        "Enable offloading support (via the non-standard `-stdpar`) for
+         Clang/LLVM. The values are AMDGPU processors (https://www.llvm.org/docs/AMDGPUUsage.html#processors)
+         which will be passed in via `--offload-arch=` argument.
+
+         Example values are:
+           gfx906  - Compile for Vega20 GPUs
+           gfx908  - Compile for CDNA1 GPUs
+           gfx90a  - Compile for CDNA2 GPUs
+           gfx1030 - Compile for RDNA2 NV21 GPUs
+           gfx1100 - Compile for RDNA3 NV31 GPUs"
+        "")
 
 macro(setup)
     set(CMAKE_CXX_STANDARD 17)
@@ -50,4 +62,13 @@ macro(setup)
         register_definitions(USE_ONEDPL)
         register_link_library(oneDPL)
     endif ()
+    if (CLANG_OFFLOAD)
+        set(CLANG_FLAGS --hipstdpar --hipstdpar-path=${CLANG_STDPAR_PATH} --offload-arch=${CLANG_OFFLOAD})
+        if (NOT CLANG_OFFLOAD MATCHES "^gfx9")
+            list(APPEND CLANG_FLAGS --hipstdpar-interpose-alloc)
+        endif ()
+
+        register_append_cxx_flags(ANY ${CLANG_FLAGS})
+        register_append_link_flags(--hipstdpar)
+    endif ()
 endmacro()
diff --git a/src/std-indices/model.cmake b/src/std-indices/model.cmake
index 60ef575..bcba9f5 100644
--- a/src/std-indices/model.cmake
+++ b/src/std-indices/model.cmake
@@ -34,6 +34,18 @@ register_flag_optional(USE_ONEDPL
           DPCPP  - Implements policies through SYCL2020.
                    This requires the DPC++ compiler (other SYCL compilers are untested), required SYCL flags are added automatically."
         "OFF")
+register_flag_optional(CLANG_OFFLOAD
+        "Enable offloading support (via the non-standard `-stdpar`) for
+         Clang/LLVM. The values are AMDGPU processors (https://www.llvm.org/docs/AMDGPUUsage.html#processors)
+         which will be passed in via `--offload-arch=` argument.
+
+         Example values are:
+           gfx906  - Compile for Vega20 GPUs
+           gfx908  - Compile for CDNA1 GPUs
+           gfx90a  - Compile for CDNA2 GPUs
+           gfx1030 - Compile for RDNA2 NV21 GPUs
+           gfx1100 - Compile for RDNA3 NV31 GPUs"
+        "")
 
 macro(setup)
     set(CMAKE_CXX_STANDARD 17)
@@ -50,4 +62,13 @@ macro(setup)
         register_definitions(USE_ONEDPL)
         register_link_library(oneDPL)
     endif ()
+    if (CLANG_OFFLOAD)
+        set(CLANG_FLAGS --hipstdpar --hipstdpar-path=${CLANG_STDPAR_PATH} --offload-arch=${CLANG_OFFLOAD})
+        if (NOT CLANG_OFFLOAD MATCHES "^gfx9")
+            list(APPEND CLANG_FLAGS --hipstdpar-interpose-alloc)
+        endif ()
+
+        register_append_cxx_flags(ANY ${CLANG_FLAGS})
+        register_append_link_flags(--hipstdpar)
+    endif ()
 endmacro()
