diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6116654..9b555bb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -50,6 +50,13 @@ if(${CMAKE_CXX_COMPILER_ID} STREQUAL GNU)
     #set(CMAKE_EXE_LINKER_FLAGS "-ltbb")
 elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL Clang)
     set(CMAKE_CXX_FLAGS "-std=c++17 -O3 -Wall -Wextra -Wpedantic")
+    if(DEFINED STDPAR_TARGET)
+        string(APPEND CMAKE_CXX_FLAGS " --hipstdpar --hipstdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET}")
+        if(NOT STDPAR_TARGET MATCHES "^gfx9")
+            string(APPEND CMAKE_CXX_FLAGS " --hipstdpar-interpose-alloc")
+        endif ()
+        set(CMAKE_EXE_LINKER_FLAGS "--hipstdpar")
+    endif()
     set (USE_TBB ON)
     #set(CMAKE_EXE_LINKER_FLAGS "-ltbb")
 elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL Intel)
