diff --git a/CMakeLists.txt b/CMakeLists.txt
index e691a7c..16461cb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -21,6 +21,13 @@ if(PROGRAMMING_MODEL STREQUAL "KOKKOS")
     endif()
 endif()
 
+if(PROGRAMMING_MODEL STREQUAL "STDPAR")
+    if(BACKEND STREQUAL "CLANG")
+        set(CMAKE_CXX_STANDARD 17)
+        set(CMAKE_CXX_STANDARD_REQUIRED ON)
+    endif()
+endif()
+
 add_subdirectory(lib)
 
 include(CTest)
diff --git a/lib/stdpar/CMakeLists.txt b/lib/stdpar/CMakeLists.txt
index 0eecba7..a0ac230 100644
--- a/lib/stdpar/CMakeLists.txt
+++ b/lib/stdpar/CMakeLists.txt
@@ -12,8 +12,27 @@ elseif(BACKEND STREQUAL "OPENMP")
     target_link_libraries(math_lib INTERFACE std::mdspan OpenMP::OpenMP_CXX fftw3)
     target_compile_definitions(math_lib INTERFACE ENABLE_OPENMP ENABLE_STDPAR)
     target_compile_options(math_lib INTERFACE -O3)
+elseif(BACKEND STREQUAL "CLANG")
+    if(DEFINED STDPAR_PATH)
+        find_package(rocblas REQUIRED)
+        find_package(rocfft REQUIRED)
+        target_link_libraries(math_lib INTERFACE std::mdspan roc::rocblas roc::rocfft)
+        target_compile_definitions(math_lib INTERFACE ENABLE_HIP ENABLE_STDPAR)
+        target_compile_options(math_lib INTERFACE --hipstdpar --hipstdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET})
+        if(NOT STDPAR_TARGET MATCHES "^gfx9")
+            target_compile_options(math_lib INTERFACE --hipstdpar-interpose-alloc)
+        endif()
+        target_link_options(math_lib INTERFACE --hipstdpar)
+    else()
+        find_package(FFTW REQUIRED COMPONENTS DOUBLE_LIB)
+        target_link_libraries(math_lib INTERFACE std::mdspan fftw3)
+        target_compile_definitions(math_lib INTERFACE ENABLE_STDPAR)
+        target_compile_options(math_lib INTERFACE -march=native -fopenmp)
+        target_link_libraries(math_lib INTERFACE tbb)
+        target_link_options(math_lib INTERFACE -fopenmp)
+    endif()
 else()
-    message(FATAL_ERROR "No parallel backend specified. One of CUDA, and OPENMP must be On.")
+    message(FATAL_ERROR "No parallel backend specified. One of CUDA, OPENMP or CLANG must be On.")
 endif()
 
 if(${CMAKE_CXX_COMPILER_ID} STREQUAL NVHPC)
diff --git a/lib/stdpar/FFT.hpp b/lib/stdpar/FFT.hpp
index a1a5d20..98fc915 100644
--- a/lib/stdpar/FFT.hpp
+++ b/lib/stdpar/FFT.hpp
@@ -3,6 +3,8 @@
 
 #if defined( _NVHPC_STDPAR_GPU )
   #include "../Cuda_FFT.hpp"
+#elif defined( __HIPSTDPAR__ )
+  #include "../HIP_FFT.hpp"
 #else
   #include "../OpenMP_FFT.hpp"
 #endif
diff --git a/lib/stdpar/Transpose.hpp b/lib/stdpar/Transpose.hpp
index 70e6044..6177794 100644
--- a/lib/stdpar/Transpose.hpp
+++ b/lib/stdpar/Transpose.hpp
@@ -3,6 +3,8 @@
 
 #if defined( _NVHPC_STDPAR_GPU )
   #include "../Cuda_Transpose.hpp"
+#elif defined( __HIPSTDPAR__ )
+  #include "../HIP_Transpose.hpp"
 #else
   #include "../OpenMP_Transpose.hpp"
 #endif
diff --git a/miniapps/heat3d/stdpar/CMakeLists.txt b/miniapps/heat3d/stdpar/CMakeLists.txt
index 1426bde..f8cff00 100644
--- a/miniapps/heat3d/stdpar/CMakeLists.txt
+++ b/miniapps/heat3d/stdpar/CMakeLists.txt
@@ -15,8 +15,19 @@ if(BACKEND STREQUAL "CUDA")
     target_link_options(heat3d PUBLIC -stdpar=gpu)
 elseif(BACKEND STREQUAL "OPENMP")
     target_compile_options(heat3d PUBLIC -stdpar=multicore)
+elseif(BACKEND STREQUAL "CLANG")
+    if(DEFINED STDPAR_PATH)
+        target_compile_options(heat3d PUBLIC --hipstdpar --hipstdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET} --save-temps --hipstdpar-interpose-alloc)
+        if(NOT STDPAR_TARGET MATCHES "^gfx9")
+            target_compile_options(heat3d PUBLIC --hipstdpar-interpose-alloc)
+        endif()
+        target_link_options(heat3d PUBLIC --hipstdpar)
+    else()
+        target_compile_options(heat3d PUBLIC -fopenmp-simd -march=native)
+        target_link_libraries(heat3d PUBLIC tbb)
+    endif()
 else()
-    message(FATAL_ERROR "No parallel backend specified. One of CUDA, and OPENMP must be On.")
+    message(FATAL_ERROR "No parallel backend specified. One of CUDA, OPENMP or CLANG must be On.")
 endif()
 
 # Macro
diff --git a/miniapps/heat3d/stdpar/Init.hpp b/miniapps/heat3d/stdpar/Init.hpp
index 84c74f2..40fa726 100644
--- a/miniapps/heat3d/stdpar/Init.hpp
+++ b/miniapps/heat3d/stdpar/Init.hpp
@@ -116,6 +116,8 @@ void performance(Config &conf, double seconds) {
     std::string backend = "CUDA";
   #elif defined( ENABLE_OPENMP )
     std::string backend = "OPENMP";
+  #elif defined( __HIPSTDPAR__ )
+    std::string backend = "CLANG";
   #else
     std::string backend = "OPENMP";
   #endif
diff --git a/miniapps/heat3d/stdpar/types.hpp b/miniapps/heat3d/stdpar/types.hpp
index ae88da4..aa3d413 100644
--- a/miniapps/heat3d/stdpar/types.hpp
+++ b/miniapps/heat3d/stdpar/types.hpp
@@ -7,7 +7,7 @@
 
 namespace stdex = std::experimental;
 
-#if defined( _NVHPC_STDPAR_GPU )
+#if defined( _NVHPC_STDPAR_GPU ) || defined( __HIPSTDPAR__ )
   #define SIMD_LOOP
   #define SIMD_WIDTH 1
   using default_layout = stdex::layout_left;
diff --git a/miniapps/heat3d_mpi/stdpar/CMakeLists.txt b/miniapps/heat3d_mpi/stdpar/CMakeLists.txt
index 1495637..52bac91 100644
--- a/miniapps/heat3d_mpi/stdpar/CMakeLists.txt
+++ b/miniapps/heat3d_mpi/stdpar/CMakeLists.txt
@@ -18,8 +18,19 @@ if(BACKEND STREQUAL "CUDA")
     target_link_options(heat3d_mpi PUBLIC -stdpar=gpu)
 elseif(BACKEND STREQUAL "OPENMP")
     target_compile_options(heat3d_mpi PUBLIC -stdpar=multicore)
+elseif(BACKEND STREQUAL "CLANG")
+    if(DEFINED STDPAR_PATH)
+        target_compile_options(heat3d_mpi PUBLIC --hipstdpar --hipstdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET})
+        if(NOT STDPAR_TARGET MATCHES "^gfx9")
+            target_compile_options(heat3d_mpi PUBLIC --hipstdpar-interpose-alloc)
+        endif()
+        target_link_options(heat3d_mpi PUBLIC --hipstdpar)
+    else()
+        target_compile_options(heat3d_mpi PUBLIC -fopenmp-simd -march=native)
+        target_link_libraries(heat3d_mpi PUBLIC tbb)
+    endif()
 else()
-    message(FATAL_ERROR "No parallel backend specified. One of CUDA, and OPENMP must be On.")
+    message(FATAL_ERROR "No parallel backend specified. One of CUDA, OPENMP or CLANG must be On.")
 endif()
 
 # Macro
diff --git a/miniapps/heat3d_mpi/stdpar/Init.hpp b/miniapps/heat3d_mpi/stdpar/Init.hpp
index f51beda..49dd286 100644
--- a/miniapps/heat3d_mpi/stdpar/Init.hpp
+++ b/miniapps/heat3d_mpi/stdpar/Init.hpp
@@ -201,7 +201,7 @@ void performance(Config &conf, Comm &comm, double seconds) {
   double GFlops = static_cast<double>(n) * size * static_cast<double>(conf.nbiter) * 9 / 1.e9;
 
   std::string backend = "STDPAR";
-  #if defined( _NVHPC_STDPAR_GPU )
+  #if defined( _NVHPC_STDPAR_GPU ) || defined( __HIPSTDPAR__ )
     std::string arch = "GPU";
   #else
     std::string arch = "CPU";
diff --git a/miniapps/heat3d_mpi/stdpar/Types.hpp b/miniapps/heat3d_mpi/stdpar/Types.hpp
index 249876d..85b923f 100644
--- a/miniapps/heat3d_mpi/stdpar/Types.hpp
+++ b/miniapps/heat3d_mpi/stdpar/Types.hpp
@@ -7,7 +7,7 @@
 
 namespace stdex = std::experimental;
 
-#if defined( _NVHPC_STDPAR_GPU )
+#if defined( _NVHPC_STDPAR_GPU ) || defined( __HIPSTDPAR__ )
   #define SIMD_LOOP
   #define SIMD_WIDTH 1
   using default_layout = stdex::layout_left;
diff --git a/miniapps/vlp4d/stdpar/CMakeLists.txt b/miniapps/vlp4d/stdpar/CMakeLists.txt
index ca5d616..c82d7ad 100644
--- a/miniapps/vlp4d/stdpar/CMakeLists.txt
+++ b/miniapps/vlp4d/stdpar/CMakeLists.txt
@@ -8,6 +8,21 @@ set(BACKEND AUTO CACHE STRING "CHOICE OF PARALLEL BACKEND")
 if(BACKEND STREQUAL "CUDA")
     target_compile_options(vlp4d PUBLIC -O3 -stdpar=gpu)
     target_link_options(vlp4d PUBLIC -stdpar=gpu -cudalib=cufft)
+elseif(BACKEND STREQUAL "CLANG")
+    if(DEFINED STDPAR_PATH)
+        find_package(rocfft REQUIRED)
+        target_compile_options(vlp4d PUBLIC --hipstdpar --hipstdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET})
+        if(NOT STDPAR_TARGET MATCHES "^gfx9")
+            target_compile_options(vlp4d PUBLIC --hipstdpar-interpose-alloc)
+        endif()
+        target_link_options(vlp4d PUBLIC --hipstdpar)
+        target_link_libraries(vlp4d PUBLIC roc::rocfft)
+    else()
+        target_compile_definitions(vlp4d PUBLIC SIMD)
+        target_compile_options(vlp4d PUBLIC -fopenmp -march=native)
+        target_link_libraries(vlp4d PUBLIC tbb)
+        target_link_options(vlp4d PUBLIC -fopenmp)
+    endif()
 else()
     target_compile_options(vlp4d PUBLIC -O3 -stdpar=multicore -mp)
     target_compile_definitions(vlp4d PUBLIC SIMD)
diff --git a/miniapps/vlp4d/stdpar/diags.hpp b/miniapps/vlp4d/stdpar/diags.hpp
index 8be727a..999a9b3 100644
--- a/miniapps/vlp4d/stdpar/diags.hpp
+++ b/miniapps/vlp4d/stdpar/diags.hpp
@@ -13,6 +13,9 @@ private:
 
 public:
   Diags(Config *conf);
+  #if defined(__HIPSTDPAR__)
+    Diags(const Diags&) = delete;
+  #endif
   virtual ~Diags();
 
   void compute(Config *conf, Efield *ef, int iter);
diff --git a/miniapps/vlp4d/stdpar/efield.hpp b/miniapps/vlp4d/stdpar/efield.hpp
index 590f138..878f518 100644
--- a/miniapps/vlp4d/stdpar/efield.hpp
+++ b/miniapps/vlp4d/stdpar/efield.hpp
@@ -26,6 +26,9 @@ private:
 
 public:
   Efield(Config *conf, shape_nd<2> dim);
+  #if defined(__HIPSTDPAR__)
+    Efield(const Efield&) = delete;
+  #endif
   virtual ~Efield();
 
   void solve_poisson_fftw(double xmax, double ymax);
diff --git a/miniapps/vlp4d/stdpar/types.hpp b/miniapps/vlp4d/stdpar/types.hpp
index 1b29915..8284f0b 100644
--- a/miniapps/vlp4d/stdpar/types.hpp
+++ b/miniapps/vlp4d/stdpar/types.hpp
@@ -9,7 +9,7 @@
 
 namespace stdex = std::experimental;
 
-#if defined( _NVHPC_STDPAR_GPU )
+#if defined( _NVHPC_STDPAR_GPU ) || defined(__HIPSTDPAR__)
   #define SIMD_LOOP
   #define SIMD_WIDTH 1
   using default_layout = stdex::layout_left;
diff --git a/tests/stdpar/CMakeLists.txt b/tests/stdpar/CMakeLists.txt
index 1429cb8..2de9d73 100644
--- a/tests/stdpar/CMakeLists.txt
+++ b/tests/stdpar/CMakeLists.txt
@@ -11,6 +11,21 @@ if(BACKEND STREQUAL "CUDA")
     target_link_options(google_tests PUBLIC -stdpar=gpu -cudalib=cufft,cublas)
 elseif(BACKEND STREQUAL "OPENMP")
     target_compile_options(google_tests PUBLIC -O3 -stdpar=multicore -mp)
+elseif(BACKEND STREQUAL "CLANG")
+    if(DEFINED STDPAR_PATH)
+        find_package(rocblas REQUIRED)
+        find_package(rocfft REQUIRED)
+        target_compile_options(google_tests PUBLIC --hipstdpar --hipstdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET})
+        if(NOT STDPAR_TARGET MATCHES "^gfx9")
+            target_compile_options(google_tests PUBLIC --hipstdpar-interpose-alloc)
+        endif()
+        target_link_libraries(google_tests PUBLIC roc::rocblas roc::rocfft)
+        target_link_options(google_tests PUBLIC --hipstdpar)
+    else()
+        target_compile_options(google_tests PUBLIC -march=native -fopenmp)
+        target_link_options(google_tests PUBLIC -fopenmp)
+        target_link_libraries(google_tests PUBLIC tbb)
+    endif()
 else()
     message(FATAL_ERROR "No parallel backend specified. One of CUDA, and OPENMP must be On.")
 endif()
diff --git a/tests/stdpar/Types.hpp b/tests/stdpar/Types.hpp
index 3344d9c..f3e3413 100644
--- a/tests/stdpar/Types.hpp
+++ b/tests/stdpar/Types.hpp
@@ -9,7 +9,7 @@
 
 namespace stdex = std::experimental;
 
-#if defined( _NVHPC_STDPAR_GPU )
+#if defined( _NVHPC_STDPAR_GPU ) || defined ( __HIPSTDPAR__ )
   #define SIMD_LOOP
   #define SIMD_WIDTH 1
   using default_layout = stdex::layout_left;
diff --git a/tests/stdpar/test_view.cpp b/tests/stdpar/test_view.cpp
index b5935a5..bd4f77d 100644
--- a/tests/stdpar/test_view.cpp
+++ b/tests/stdpar/test_view.cpp
@@ -28,7 +28,8 @@ void set_inside_function(RealView2D shallow_copy_to_a_View) {
     });
 }
 
-void test_copy_constructor() {
+#if !defined(__HIPSTDPAR__) && !defined(__GNUC__)
+  void test_copy_constructor() {
     RealView2D simple("simple", 16, 16);
     RealView2D reference("reference", 16, 16);
 
@@ -56,7 +57,8 @@ void test_copy_constructor() {
         ASSERT_NE( simple(i, j), 0.0 ); // Just to make sure simple has some value
       }
     }
-}
+  }
+#endif
 
 void test_assignment_operator() {
   // [NOTE] Do not recommend to use assignement opertor
@@ -236,9 +238,11 @@ TEST( VIEW, DEFAULT_CONSTRUCTOR ) {
   RealView2D offset_view_int("offset_view", std::array<int, 2>{3, 4}, std::array<int, 2>{-1, -1}); // Offset view
 }
 
-TEST( VIEW, COPY_CONSTRUCTOR ) {
+#if !defined(__HIPSTDPAR__) && !defined(__GNUC__)
+  TEST( VIEW, COPY_CONSTRUCTOR ) {
     test_copy_constructor();
-}
+  }
+#endif
 
 TEST( VIEW, ASSIGN ) {
   test_assignment_operator();
