diff --git a/stdpar/himenoBMTxps.cpp b/stdpar/himenoBMTxps.cpp
index 8a95e7d..1e77079 100644
--- a/stdpar/himenoBMTxps.cpp
+++ b/stdpar/himenoBMTxps.cpp
@@ -119,9 +119,11 @@ public:
       using iterator_category = std::random_access_iterator_tag;
       explicit iterator(N _num = 0) : num(_num) {}
 
-      iterator& operator++() { num++; return *this; }
+      iterator& operator++() { ++num; return *this; }
+      iterator& operator--() { --num; return *this; }
       iterator operator++(int) { iterator retval = *this; ++(*this); return retval; }
       iterator operator+(const value_type v) const { return iterator(num + v); }
+      iterator& operator+=(const value_type v) { num += v; return *this; }
 
       bool operator==(iterator other) const { return num == other.num; }
       bool operator!=(iterator other) const { return *this != other; }
@@ -346,4 +348,8 @@ double second()
 // nvc++ -O3 -std=c++17 -stdpar=gpu -DLARGE himenoBMTxps.cpp -o himenoBMTxps && ./himenoBMTxps
 // nvc++ -O3 -std=c++17 -stdpar=gpu -DMIDDLE himenoBMTxps.cpp -o himenoBMTxps && ./himenoBMTxps
 // nvc++ -O3 -std=c++17 -stdpar=gpu -DSMALL himenoBMTxps.cpp -o himenoBMTxps && ./himenoBMTxps
+// OR
+// clang++ -O3 -std=c++17 --hipstdpar --hipstdpar-path=PATH_TO_FORWARDING_HEADER_ROOT --offload-arch=YOUR_TARGET -DLARGE himenoBMTxps.cpp -o himenoBMTxps && ./himenoBMTxps
+// clang++ -O3 -std=c++17 --hipstdpar --hipstdpar-path=PATH_TO_FORWARDING_HEADER_ROOT --offload-arch=YOUR_TARGET -DMIDDLE himenoBMTxps.cpp -o himenoBMTxps && ./himenoBMTxps
+// clang++ -O3 -std=c++17 --hipstdpar --hipstdpar-path=PATH_TO_FORWARDING_HEADER_ROOT --offload-arch=YOUR_TARGET -DSMALL himenoBMTxps.cpp -o himenoBMTxps && ./himenoBMTxps
 
