diff --git a/stdpar/Makefile b/stdpar/Makefile
index 96b6d9d..82929ee 100644
--- a/stdpar/Makefile
+++ b/stdpar/Makefile
@@ -32,9 +32,19 @@ tsp_pgi_stdpar_gpu: tsp.cpp tsp_utils.h route_cost.h route_iterator.h counting_i
 tsp_dpcpp_stdpar_cpu: tsp_oneapi.cpp tsp_utils.h route_cost.h route_iterator.h counting_iterator.h
 	dpcpp -O3 -std=c++17 -ferror-limit=3 $< -o $@ -L${TBB_LIB} -ltbb
 
-all: tsp_gnu_stdpar_cpu tsp_pgi_stdpar_cpu
+#
+# Clang / LLVM Standard Parallelism using HIP on the ROCM toolchain
+#
+tsp_clang_stdpar_cpu: tsp.cpp tsp_utils.h route_cost.h route_iterator.h counting_iterator.h
+	${STDPAR_CXX} -O3 -march=native -std=c++17 -fopenmp-simd $< -o $@ -L${TBB_LIB} -ltbb
+
+tsp_clang_stdpar_gpu: tsp.cpp tsp_utils.h route_cost.h route_iterator.h counting_iterator.h
+	${STDPAR_CXX} -O3 -march=native -std=c++17 -stdpar --stdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET} $< -o $@
+
+all: tsp_gnu_stdpar_cpu tsp_pgi_stdpar_cpu tsp_clang_stdpar_cpu
 
 clean:
 	\rm -f tsp_gnu_stdpar_cpu tsp_gnu10_stdpar_cpu tsp_gnu11_stdpar_cpu
 	\rm -f tsp_pgi_stdpar_cpu tsp_pgi_stdpar_gpu
 	\rm -f tsp_dpcpp_stdpar_cpu
+	\rm -f tsp_clang_stdpar_cpu tsp_clang_stdpar_gpu
\ No newline at end of file
diff --git a/stdpar/tsp.cpp b/stdpar/tsp.cpp
index 45864ce..fa52ce3 100644
--- a/stdpar/tsp.cpp
+++ b/stdpar/tsp.cpp
@@ -86,7 +86,7 @@ route_cost find_best_route(int const* distances)
 {
 
   return
-    std::transform_reduce(std::execution::par,
+    std::transform_reduce(std::execution::par_unseq,
                           counting_iterator(0),
                           counting_iterator(factorial(N)),
                           route_cost(),
