diff --git a/build/.gitignore b/build/.gitignore
deleted file mode 100644
index 86d0cb2..0000000
--- a/build/.gitignore
+++ /dev/null
@@ -1,4 +0,0 @@
-# Ignore everything in this directory
-*
-# Except this file
-!.gitignore
\ No newline at end of file
diff --git a/examples/80days/rayleighTaylor3D/CMakeLists.txt b/examples/80days/rayleighTaylor3D/CMakeLists.txt
index d304893..0cf2664 100644
--- a/examples/80days/rayleighTaylor3D/CMakeLists.txt
+++ b/examples/80days/rayleighTaylor3D/CMakeLists.txt
@@ -49,6 +49,13 @@ if(${CMAKE_CXX_COMPILER_ID} STREQUAL GNU)
 elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL Clang)
     message("Clang.")
     set(CMAKE_CXX_FLAGS "-std=c++17 -fopenmp -Wall -Wnon-virtual-dtor")
+    if (DEFINED CLANG_STDPAR)
+        string(APPEND CMAKE_CXX_FLAGS " --hipstdpar --hipstdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET}")
+        if (NOT STDPAR_TARGET MATCHES "^gfx9")
+            string(APPEND CMAKE_CXX_FLAGS " --hipstdpar-interpose-alloc")
+        endif ()
+        set(CMAKE_EXE_LINKER_FLAGS --hipstdpar)
+    endif ()
     set(CMAKE_CXX_FLAGS_RELEASE "-O3 -DNDEBUG")
     set(CMAKE_CXX_FLAGS_DEBUG "-g -DPLB_DEBUG -O1")
 elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL PGI) # nvc++ is based on the PGI compiler
diff --git a/examples/80days/sandstone/CMakeLists.txt b/examples/80days/sandstone/CMakeLists.txt
index 823b3b7..43a0ea1 100644
--- a/examples/80days/sandstone/CMakeLists.txt
+++ b/examples/80days/sandstone/CMakeLists.txt
@@ -49,6 +49,13 @@ if(${CMAKE_CXX_COMPILER_ID} STREQUAL GNU)
 elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL Clang)
     message("Clang.")
     set(CMAKE_CXX_FLAGS "-std=c++17 -fopenmp -Wall -Wnon-virtual-dtor")
+    if (DEFINED CLANG_STDPAR)
+        string(APPEND CMAKE_CXX_FLAGS " --hipstdpar --hipstdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET}")
+        if (NOT STDPAR_TARGET MATCHES "^gfx9")
+            string(APPEND CMAKE_CXX_FLAGS " --hipstdpar-interpose-alloc")
+        endif ()
+        set(CMAKE_EXE_LINKER_FLAGS --hipstdpar)
+    endif ()
     set(CMAKE_CXX_FLAGS_RELEASE "-O3 -DNDEBUG")
     set(CMAKE_CXX_FLAGS_DEBUG "-g -DPLB_DEBUG -O1")
 elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL PGI) # nvc++ is based on the PGI compiler
diff --git a/examples/80days/tgv3d/CMakeLists.txt b/examples/80days/tgv3d/CMakeLists.txt
index d37e5ed..3a1d2ba 100644
--- a/examples/80days/tgv3d/CMakeLists.txt
+++ b/examples/80days/tgv3d/CMakeLists.txt
@@ -49,6 +49,13 @@ if(${CMAKE_CXX_COMPILER_ID} STREQUAL GNU)
 elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL Clang)
     message("Clang.")
     set(CMAKE_CXX_FLAGS "-std=c++17 -fopenmp -Wall -Wnon-virtual-dtor")
+    if (DEFINED CLANG_STDPAR)
+        string(APPEND CMAKE_CXX_FLAGS " --hipstdpar --hipstdpar-path=${STDPAR_PATH} --offload-arch=${STDPAR_TARGET}")
+        if (NOT STDPAR_TARGET MATCHES "^gfx9")
+            string(APPEND CMAKE_CXX_FLAGS " --hipstdpar-interpose-alloc")
+        endif ()
+        set(CMAKE_EXE_LINKER_FLAGS -hipstdpar)
+    endif ()
     set(CMAKE_CXX_FLAGS_RELEASE "-O3 -DNDEBUG")
     set(CMAKE_CXX_FLAGS_DEBUG "-g -DPLB_DEBUG -O1")
 elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL PGI) # nvc++ is based on the PGI compiler
